pfcpiface/bess.go

```diff
@@ -342,6 +343,202 @@ func (b *bess) summaryLatencyJitter(uc *upfCollector, ch chan<- prometheus.Metri
+	}
+	// Determine new flag
+	var newFlag pb.BufferFlag
+	if oldFlag == pb.BufferFlag_FLAG_VALUE_A {
+		newFlag = pb.BufferFlag_FLAG_VALUE_B
+	} else {
+		newFlag = pb.BufferFlag_FLAG_VALUE_A
+	}
+	log.Traceln("old flag:", oldFlag, "new flag:", newFlag)
+	// Set new flag
+	req := &pb.DoubleBufferCommandSetNewFlagValueArg{NewFlag: newFlag}
+	any, err := anypb.New(req)
+	if err != nil {
+		log.Errorln("Error marshalling request", req, err)
+		return
+	}
+	resp, err := b.client.ModuleCommand(
+		ctx, &pb.CommandRequest{
+			Name: "flag",
+			Cmd:  "set",
+			Arg:  any,
+		},
+	)
+	if err != nil {
+		log.Errorln("flag set failed!:", err)
+		return
+	}
+	flagSetResp := &pb.DoubleBufferCommandSetNewFlagValueResponse{}
+	if err = resp.Data.UnmarshalTo(flagSetResp); err != nil {
+		log.Errorln(err)
+		return
+	}
+	durationNs = flagSetResp.ObservationDurationNs
+	return
+}
+
+func (b *bess) readMeasurement(
+	ctx context.Context, module string, flag pb.BufferFlag, clear bool,
+) (stats pb.QosMeasureReadResponse, err error) {
+	req := &pb.QosMeasureCommandReadArg{Flag: flag, Clear: clear}
+	any, err := anypb.New(req)
+	if err != nil {
+		log.Errorln("Error marshalling request", req, err)
+		return
+	}
+	resp, err := b.client.ModuleCommand(
+		ctx, &pb.CommandRequest{
+			Name: module,
+			Cmd:  "read",
+			Arg:  any,
+		},
+	)
+	if err != nil {
+		log.Errorln(module, "read failed!:", err)
+		return
+	}
+	if err = resp.Data.UnmarshalTo(&stats); err != nil {
+		log.Errorln(err)
+		return
+	}
+	return
+}
+
+func (b *bess) sessionStats(uc *upfCollector, ch chan<- prometheus.Metric) (err error) {
+	// Clearing table data with large tables is slow, let's wait for a little longer since this is
+	// non-blocking anyway.
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second*5)
+	defer cancel()
+	// Flip flag, automatically waits for pipeline to flush packets with previous mark.
+	observationDurationNs, oldFlag, err := b.flipFlag(ctx)
+	if err != nil {
+		return err
+	}
+	// Read stats from the now inactive side, and clear if needed.
+	qosStatsInResp, err := b.readMeasurement(ctx, "preQosMeasure", oldFlag, true)
+	if err != nil {
+		log.Errorln("preQosMeasure read failed!:", err)
+		return
+	}
+	postDlQosStatsResp, err := b.readMeasurement(ctx, "postDLQoSMeasure", oldFlag, true)
+	if err != nil {
+		log.Errorln("postDLQoSMeasure read failed!:", err)
+		return
+	}
+	postUlQosStatsResp, err := b.readMeasurement(ctx, "postULQoSMeasure", oldFlag, true)
+	if err != nil {
+		log.Errorln("postULQoSMeasure read failed!:", err)
+		return
+	}
+
+	// Prepare prometheus stats.
+	createStats := func(pre, post *pb.QosMeasureReadResponse_Statistic, ch chan<- prometheus.Metric) {
+		dropRate := 1 - (float64(post.TotalPackets) / float64(pre.TotalPackets))
+		bitsPerSecond := float64(post.TotalBytes*8) / (float64(observationDurationNs) / (1000 * 1000 * 1000))
+		packetsPerSecond := float64(post.TotalPackets) / (float64(observationDurationNs) / (1000 * 1000 * 1000))
```

Comment:
Should we export absolute numbers of {total, dropped}_{bytes, packets} and then let users consuming them via prometheus/grafana calculate the rates and drop %-tages etc., if need be?